clang/labs/lab1/polozov_vladislav/WarningForDeprecated.cpp

```diff
@@ -0,0 +1,60 @@
+#include "clang/AST/ASTConsumer.h"
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/Frontend/CompilerInstance.h"
+#include "clang/Frontend/FrontendPluginRegistry.h"
+#include "llvm/Support/raw_ostream.h"
+#include <algorithm>
+using namespace clang;
+
+class WarningDeprecatedConsumer : public ASTConsumer {
+  CompilerInstance &Instance;
+
+public:
+  WarningDeprecatedConsumer(CompilerInstance &Instance) : Instance(Instance) {}
+
+  void HandleTranslationUnit(ASTContext &context) override {
+
+    struct Visitor : public RecursiveASTVisitor<Visitor> {
+      ASTContext *Context;
+      Visitor(ASTContext *_Context) : Context(_Context) {}
+      bool VisitFunctionDecl(FunctionDecl *FD) {
+        std::string want_find = "deprecated";
+        std::string name = FD->getNameInfo().getAsString();
+        std::transform(name.begin(),name.end(),name.begin(), ::tolower);
+        if (name.find(want_find) != std::string::npos) {
+          DiagnosticsEngine &Diags = Context->getDiagnostics();
+          unsigned DiagID = Diags.getCustomDiagID(
+              DiagnosticsEngine::Warning, "find 'deprecated' in function name");
+          Diags.Report(FD->getLocation(), DiagID)
+              << FD->getNameInfo().getAsString();
+        }
+        return true;
+      }
+    };
+    Visitor MyVisitor(&Instance.getASTContext());
+    MyVisitor.TraverseDecl(context.getTranslationUnitDecl());
```

Comment:
Please, rename the visitor

Reply:
Better to add more descriptive name to your visitor. It would simplify code reading and a search in the whole clang's code. Like `DeprWarnPrintVisitor`.